src/legacy/ui/public/saved_objects/components/saved_object_finder.tsx

```diff
@@ -197,124 +225,291 @@ class SavedObjectFinder extends React.Component<SavedObjectFinderProps, SavedObj
   };
 
   private fetchItems = () => {
     this.setState(
       {
         isFetchingItems: true,
       },
-      this.debouncedFetch.bind(null, this.state.filter)
+      this.debouncedFetch.bind(null, this.state.query)
     );
   };
 
+  private getAvailableSavedObjectMetaData() {
+    const typesInItems = new Set<string>();
+    this.state.items.forEach(item => {
+      typesInItems.add(item.type);
+    });
+    return this.props.savedObjectMetaData.filter(metaData => typesInItems.has(metaData.type));
+  }
+
+  private getSortOptions() {
+    const sortOptions = [
+      <EuiContextMenuItem
+        key="asc"
+        icon={
+          this.state.sortDirection === 'asc' ||
+          (this.state.query === '' && this.state.sortDirection !== 'desc')
+            ? 'check'
+            : 'empty'
+        }
+        onClick={() => {
+          this.setState({
+            sortDirection: 'asc',
+          });
+        }}
+      >
+        {i18n.translate('common.ui.savedObjects.finder.sortAsc', {
+          defaultMessage: 'Ascending',
+        })}
+      </EuiContextMenuItem>,
+      <EuiContextMenuItem
+        key="desc"
+        icon={this.state.sortDirection === 'desc' ? 'check' : 'empty'}
+        onClick={() => {
+          this.setState({
+            sortDirection: 'desc',
+          });
+        }}
+      >
+        {i18n.translate('common.ui.savedObjects.finder.sortDesc', {
+          defaultMessage: 'Descending',
+        })}
+      </EuiContextMenuItem>,
+    ];
+    if (this.state.query) {
+      sortOptions.push(
+        <EuiContextMenuItem
+          key="auto"
+          icon={!this.state.sortDirection ? 'check' : 'empty'}
+          onClick={() => {
+            this.setState({
+              sortDirection: undefined,
+            });
+          }}
+        >
+          {i18n.translate('common.ui.savedObjects.finder.sortAuto', {
+            defaultMessage: 'Best match',
+          })}
+        </EuiContextMenuItem>
+      );
+    }
+    return sortOptions;
+  }
+
   private renderSearchBar() {
+    const availableSavedObjectMetaData = this.getAvailableSavedObjectMetaData();
+
     return (
-      <EuiFlexGroup>
+      <EuiFlexGroup gutterSize="m">
         <EuiFlexItem grow={true}>
           <EuiFieldSearch
             placeholder={i18n.translate('common.ui.savedObjects.finder.searchPlaceholder', {
               defaultMessage: 'Search…',
             })}
             fullWidth
-            value={this.state.filter}
+            value={this.state.query}
             onChange={e => {
               this.setState(
                 {
-                  filter: e.target.value,
+                  query: e.target.value,
                 },
                 this.fetchItems
               );
             }}
             data-test-subj="savedObjectFinderSearchInput"
           />
+          {this.state.isFetchingItems && (
+            <EuiProgress size="xs" color="accent" className="savedObjectFinderLoadingBar" />
```

Comment:
The `EuiFieldSearch` actually has a `isLoading` prop you can use here instead which essentially just adds the spinner to the right of the field:
<img width="440" alt="screen shot 2019-03-08 at 10 08 28 am" src="https://user-images.githubusercontent.com/549577/54036756-69eddf80-418a-11e9-91f2-47bfad85bc0d.png">

Reply:
Ah, didn't see this, thanks.